agent/rpc/peering/replication.go

```diff
@@ -282,8 +288,136 @@ func (s *Service) handleUpsertService(
+			// (but not the node itself) if the node wasn't in the snapshot,
+			// so we do not loop over checks here.
+			// If the unusedNode gets deleted below that will also delete node checks.
+			continue
+		}
+
+		// Delete the service instance if not in the snapshot.
+		sid := csn.Service.CompoundServiceID()
+		if _, ok := snap.Nodes[csn.Node.ID].Services[sid]; !ok {
+			err := s.Backend.Apply().CatalogDeregister(&structs.DeregisterRequest{
+				Node:           csn.Node.Node,
+				ServiceID:      csn.Service.ID,
+				EnterpriseMeta: csn.Service.EnterpriseMeta,
+				PeerName:       peerName,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to deregister service: %w", err)
+			}
+
+			// When a service is deleted all associated checks also get deleted as a side effect.
+			continue
+		}
+
+		// Reconcile checks.
+		for _, chk := range csn.Checks {
+			if _, ok := snap.Nodes[csn.Node.ID].Services[sid].Checks[chk.CheckID]; !ok {
+				// Checks without a ServiceID are node checks.
+				// If the node exists but the check does not then the check was deleted.
+				if chk.ServiceID == "" {
+					// Deduplicate node checks to avoid deregistering a check multiple times.
+					tuple := nodeCheckTuple{
+						checkID: chk.CheckID,
+						node:    chk.Node,
+					}
+					deletedNodeChecks[tuple] = struct{}{}
+					continue
+				}
+
+				// If the check isn't a node check then it's a service check.
+				// Service checks that were not present can be deleted immediately because
+				// checks for a given service ID will only be attached to a single CheckServiceNode.
+				err := s.Backend.Apply().CatalogDeregister(&structs.DeregisterRequest{
+					Node:           chk.Node,
+					CheckID:        chk.CheckID,
+					EnterpriseMeta: chk.EnterpriseMeta,
+					PeerName:       peerName,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to deregister check: %w", err)
+				}
+			}
+		}
+	}
 
+	// Delete all deduplicated node checks.
+	for chk := range deletedNodeChecks {
+		err := s.Backend.Apply().CatalogDeregister(&structs.DeregisterRequest{
+			Node:           chk.node,
+			CheckID:        chk.checkID,
+			EnterpriseMeta: *structs.NodeEnterpriseMetaInPartition(sn.PartitionOrDefault()),
+			PeerName:       peerName,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to deregister check: %w", err)
+		}
+	}
+
+	// Delete any nodes that do not have any other services registered on them.
+	for node := range unusedNodes {
+		_, ns, err := s.Backend.Store().NodeServices(nil, string(node), &sn.EnterpriseMeta, peerName)
+		if err != nil {
+			return fmt.Errorf("failed to query services on node: %w", err)
+		}
+		if ns != nil && len(ns.Services) >= 1 {
+			// At least one service is still registered on this node, so we keep it.
+			continue
+		}
+
+		// All services on the node were deleted, so the node is also cleaned up.
+		err = s.Backend.Apply().CatalogDeregister(&structs.DeregisterRequest{
+			Node:     string(node),
```

Comment:
You need to specify the partition for the node using a empty-namespace, populated-partition entmeta here.

Reply:
Ah yep good catch

Reply:
I'll add some ent tests in a separate PR